mktrayicon.c

```diff
@@ -320,79 +328,92 @@ gpointer watch_fifo(gpointer argv)
 #ifdef DEBUG
 				printf("Removing onmenu handler\n");
 #endif
+				free(param);
 				break;
 			}
 
-			// This block makes sure that the parameter after 'm' is well defined
-			int commas = 0;
-			int bars = 0;
-			if (len < 3){
-				break;
-			}
-			if (param[0] == ',' || param[0] == '|'){
-				printf("Action string cannot begin with ',' or '|'. Use '\\' to escape\n");
-				break;
-			}
-			if (param[len-2] == '|' && param[len-3] != '\\'){
-				printf("'|' can't be second of last. Use '\\' to escape\n");
-				break;
+			// This block makes sure that the parameter after 'm' is ready to be processed
+			
+			char* initial = param; 
+			//we save so that we can free later even if we increase the pointer
+			
+			while (param[0] == ',' || param[0] == '|'){
+				// ignore delimiter characters in first spot
+				param++;
+				len--;
 			}
-			if (param[len-1] == ',' && param[len-2] != '\\'){
-				printf("',' can't be last. Use '\\' to escape\n");
-				break;
+			while ((param[len-1] == ',' || param[len-1] == '|') && param[len-2] != '\\'){
+				// ignore delimiters at the end too
+				// it also helps dividing the string in entries -> bars+1
+				param[len-1] = '\0';
+				len--;
 			}
-			int i = 1; 
-			while (i < len - 1){
-				if (commas > bars + 1 || bars > commas)
-					break;
-				if (param[i] == ',' && param[i-1] != '\\')
-					commas++;
-				else if (param[i] == '|' && param[i-1] != '\\')
+
+			// we can't accept 2 straight commas, as it becomes ambiguous
+			int straight = 0;
+			int bars = 0;
+			for (int i = 0; i < len; i++){
+				if (param[i] == ',' && param[i-1] != '\\'){
+					straight++;
+					if(straight == 2)
+						break;
+				}
+				else if (param[i] == '|' && param[i-1] != '\\' ){
+					straight = 0;
 					bars++;
-				i++;
+				}
 			}
-			if (commas > bars + 1 || bars > commas){
-				printf("Unbalanced ',' or '|'. Use '\\' to escape\n");
+			if (straight == 2){
+				printf("Two straight ',' found. Use '\\' to escape\n");
+				free(initial);
 				break;
 			}
 			// End of block that checks the parameter
 			
 			// Create the onmenu array which stores structs with name, action properties
-			menusize = commas;
+			menusize = bars + 1;
```

Comment:
It seems odd to me to use `bars` here. Do we want to _require_ that every menu item has a command? Label-only menu items have some useful use-cases I think.

Reply:
Entries are separated by bars. So `label1|label2|label3` will create 3 empty labels

Reply:
Ah, sorry, I though `,` was the item separator and `|` was the label/action separator! My bad.

To me, `label,|` and `,|label` should both be accepted. They both parse into two menu items. The first produces a menu item with label "label" and no action, followed by a menu item with no label or action. The second produces a menu item with no label or action, followed by a menu item with the label "label" and no action. As mentioned earlier, when we actually _walk_ the menu items to produce the context menu, we could skip parsed items with no label and warn the user about them.

Reply:
So if I understand correctly you want to have `#entries = #bars + 1` whatever the position the bars are in

Reply:
Exactly.

Reply:
Ok, given the code structure it should be relatively easy to implement. I will make a new commit when it's ready.

Reply:
Quick question. What should we do with two straight commas? Currently it's disallowed. I think we should leave it at that

Reply:
Yeah, I think two commas in one menu item should be an error.